01.fit_a_line/train.py

```diff
@@ -13,122 +13,135 @@
-    batch_size=BATCH_SIZE)
 
+def main():
 
-def train_program():
-    y = fluid.layers.data(name='y', shape=[1], dtype='float32')
+    batch_size = 20
+    train_reader = paddle.batch(
+        paddle.reader.shuffle(paddle.dataset.uci_housing.train(), buf_size=500),
+        batch_size=batch_size)
+    test_reader = paddle.batch(
+        paddle.reader.shuffle(paddle.dataset.uci_housing.test(), buf_size=500),
+        batch_size=batch_size)
 
     # feature vector of length 13
     x = fluid.layers.data(name='x', shape=[13], dtype='float32')
+    y = fluid.layers.data(name='y', shape=[1], dtype='float32')
     y_predict = fluid.layers.fc(input=x, size=1, act=None)
 
-    loss = fluid.layers.square_error_cost(input=y_predict, label=y)
-    avg_loss = fluid.layers.mean(loss)
-
-    return avg_loss
-
-
-def optimizer_program():
-    return fluid.optimizer.SGD(learning_rate=0.001)
-
-
-# can use CPU or GPU
-use_cuda = False
-place = fluid.CUDAPlace(0) if use_cuda else fluid.CPUPlace()
-
-trainer = Trainer(
-    train_func=train_program, place=place, optimizer_func=optimizer_program)
-
-feed_order = ['x', 'y']
-
-# Specify the directory to save the parameters
-params_dirname = "fit_a_line.inference.model"
-
-train_title = "Train cost"
-test_title = "Test cost"
-
-step = 0
-
-
-# event_handler prints training and testing info
-def event_handler(event):
-    global step
-    if isinstance(event, EndStepEvent):
-        if step % 10 == 0:  # record a train cost every 10 batches
-            print("%s, Step %d, Cost %f" %
-                  (train_title, step, event.metrics[0]))
-        if step % 100 == 0:  # record a test cost every 100 batches
-            test_metrics = trainer.test(
-                reader=test_reader, feed_order=feed_order)
-            print("%s, Step %d, Cost %f" % (test_title, step, test_metrics[0]))
-            if test_metrics[0] < 10.0:
-                # If the accuracy is good enough, we can stop the training.
-                print('loss is less than 10.0, stop')
-                trainer.stop()
-        step += 1
-
-    if isinstance(event, EndEpochEvent):
-        if event.epoch % 10 == 0:
-            # We can save the trained parameters for the inferences later
+    main_program = fluid.default_main_program()
+    star_program = fluid.default_startup_program()
+
+    cost = fluid.layers.square_error_cost(input=y_predict, label=y)
+    avg_loss = fluid.layers.mean(cost)
+
+    sgd_optimizer = fluid.optimizer.SGD(learning_rate=0.001)
+    sgd_optimizer.minimize(avg_loss)
+
+    test_program = main_program.clone(for_test=True)
+
+    # can use CPU or GPU
+    use_cuda = False
+    place = fluid.CUDAPlace(0) if use_cuda else fluid.CPUPlace()
+    exe = fluid.Executor(place)
+
+    # Specify the directory to save the parameters
+    params_dirname = "fit_a_line.inference.model"
+    num_epochs = 100
+
+    # For training test cost
+    def train_test(program, feeder):
+        exe_test = fluid.Executor(place)
```

Comment:
It is better to use ParallelExecutor rather than Executor for training and testing.